src/Orleans/Utils/Factory.cs

```diff
@@ -0,0 +1,36 @@
+﻿namespace Orleans
+{
+    /// <summary>
+    /// Creates an instance of <typeparamref name="TInstance"/>.
+    /// </summary>
+    /// <typeparam name="TInstance"></typeparam>
+    /// <returns></returns>
+    public delegate TInstance Factory<out TInstance>();
```

Comment:
I don't have a _strong_ opinion on this, but my preference would be to request a factory interface from DI rather than a delegate.  My reasons for this preference are:

- Factory classes can support a wider set of behaviors, including statistics tracking and state-full factories like object pools.
- Factory implementations can be constructed like any other DI type so they can have access to the container, configuration, and other application or framework classes.
- Factory interfaces can be decorated to stack application layer pooling, monitoring or other behaviors, making them very customizable by environment/scenario (test, load, production)
- Existing systems can be easily hidden behind factory interfaces, making integration easier.

Something like:

    public interface IFactory<out TType>
        where TType : class
    {
        TType Create();
    }